### PR TITLE
fix(darwin): Start observing `AVPlayerItem.status` before being assigned to `AVPlayer`

### PR DESCRIPTION
**Pull request by blaugold**
# Description

Beginning with `4.1.0` we are seeing a lot of `TimeoutException after 0:00:30.000000: Future not completed` exceptions on iOS. 

I think I have tracked it down to when the KV-Observer for `AVPlayerItem.status` is registered. Before this change this was done after setting the `AVPlayerItem` on the `AVPlayer`. The problem with this is that `AVPlayerItem` starts loading immediately after setting it on the `AVPlayer` and can finish loading before the KV-Observer is even registered. 

Now the KV-Observer is registered just after creating the `AVPlayerItem`.

The fix could have been done with less changes, but I though `WrappedMediaPlayer` could be clean up a bit.

## Checklist

- [x] The title of my PR starts with a [Conventional Commit] prefix (`fix:`, `feat:`, `docs:`, `chore:` etc).
- [x] I have read the [Contributor Guide] and followed the process outlined for submitting PRs.
- [x] I have updated/added tests for ALL new/updated/fixed functionality.
- [x] I have updated/added relevant documentation and added dartdoc comments with `///`, where necessary.
- [x] I have updated/added relevant examples in [example].

## Breaking Change

- [ ] Yes, this is a breaking change.
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/bluefireteam/audioplayers/issues
[Contributor Guide]: https://github.com/bluefireteam/audioplayers/blob/main/contributing.md#feature-requests--prs
[Conventional Commit]: https://conventionalcommits.org
[example]: https://github.com/bluefireteam/audioplayers/tree/main/packages/audioplayers/example

### CODE REVIEW

**Review by spydon — APPROVED** [review]

File: packages/audioplayers/lib/src/audioplayer.dart [hunk 1]
```diff
@@ -298,9 +298,12 @@ class AudioPlayer {
         }
       },
     );
-    await fun();
-    await preparedCompleter.future.timeout(const Duration(seconds: 30));
-    onPreparedSubscription.cancel();
+    try {
+      await fun();
+      await preparedCompleter.future.timeout(const Duration(seconds: 30));
```
Comment (Gustl22) [inline packages/audioplayers/lib/src/audioplayer.dart:303]: I think this is not the right way here: we should avoid catching any error and instead try to not emit the prepared more than once or reinitialize the onPreparedCompleter.

**Comment by Gustl22** [comment]
Hi @blaugold thank you very much for contributing! We love that :) 
Would you mind to split the refactor part in a new PR from the actual fix, otherwise the changes are hard to track, and then maybe we could add a test to avoid reverting the fix? Thank you for your help!

**Comment by Gustl22** [comment]
@blaugold little reminder, if you still working on this :)

**Comment by blaugold** [comment]
@Gustl22 Thanks for reviewing this PR and the ping. I'm still planning on addressing the comments, just haven't had the time yet :)

File: packages/audioplayers_darwin/darwin/Classes/WrappedMediaPlayer.swift [hunk 2]
```diff
@@ -43,35 +43,22 @@ class WrappedMediaPlayer {
...
-                self.observers = []
-                self.url = url
-
-                setUpPositionObserver(player)
-            }
-
-            setUpSoundCompletedObserver(player, playerItem)
             setUpPlayerItemStatusObservation(playerItem, completer, completerError)
+            let player = updatePlayer(playerItem) ?? createPlayer(playerItem)
+            setUpPositionObserver(player)
```
Comment (Gustl22) [inline packages/audioplayers_darwin/darwin/Classes/WrappedMediaPlayer.swift:60]: Setting up position observer for both creating and updating the player means that previously the position was not updated, when setting a new url? Is that right? So this is a fix then maybe? Or if not, why did it work in the first place?

File: packages/audioplayers_darwin/darwin/Classes/WrappedMediaPlayer.swift [hunk 3]
```diff
@@ -171,25 +158,38 @@ class WrappedMediaPlayer {
         return playerItem
     }
 
+    private func createPlayer( _ playerItem: AVPlayerItem) -> AVPlayer {
+        let player = AVPlayer.init(playerItem: playerItem)
+        configParameters(player: player)
+        self.player = player
```
Comment (Gustl22) [inline packages/audioplayers_darwin/darwin/Classes/WrappedMediaPlayer.swift:164]: We should either return the player or set it in the method, but not both, as it leads to confusion. I suggest just returning the player object.

File: packages/audioplayers_darwin/darwin/Classes/WrappedMediaPlayer.swift [hunk 4]
```diff
@@ -171,25 +158,38 @@ class WrappedMediaPlayer {
...
             case .failed:
+                self.eventHandler.onLog(message: "error: \(String(describing: playerItem.error))")
                 self.reset()
                 completerError?()
             default:
                 break
             }
         }
-    
-        playerItemStatusObservation?.invalidate()
```
Comment (Gustl22) [inline packages/audioplayers_darwin/darwin/Classes/WrappedMediaPlayer.swift:191]: Just as note for myself: Invalidate is now called in `reset()` for both creating and updating the player

**Review by Gustl22 — COMMENTED** [review]
I added another Pull request to better distinguish the refactor from the actual logical changes, for a better review. Hope that is ok for you.

File: packages/audioplayers_darwin/darwin/Classes/WrappedMediaPlayer.swift [hunk 5]
```diff
@@ -171,25 +158,38 @@ class WrappedMediaPlayer {
...
+        playerItemStatusObservation = playerItem.observe(\AVPlayerItem.status) { (playerItem, change) in
+            let status = playerItem.status
             self.eventHandler.onLog(message: "player status: \(status), change: \(change)")
 
             switch playerItem.status {
             case .readyToPlay:
                 self.updateDuration()
                 completer?()
             case .failed:
+                self.eventHandler.onLog(message: "error: \(String(describing: playerItem.error))")
```
Comment (Gustl22) [inline packages/audioplayers_darwin/darwin/Classes/WrappedMediaPlayer.swift:186]: The observer is already completing with an error, no need to write this to the eventHandler.

**Comment by Gustl22** [comment]
@blaugold small reminder, if you still plan to work on this. Thank you ;D

File: packages/audioplayers_darwin/darwin/Classes/WrappedMediaPlayer.swift (hunk 2 above)
Comment (Gustl22) [reply packages/audioplayers_darwin/darwin/Classes/WrappedMediaPlayer.swift:60]: Now I get it: the position update is bound to the `player` instance, where as the completedObserver is bound to the `playerItem` instance.
I would propose the following:
- We move the `createPlayer` part to the init method or/and just use `player = AVPlayer.init()`, as this is only called once, during its live time.
- We only use `updatePlayer` in `setSourceUrl`, but can be sure, that player is never nil.
- Split observers in a `positionObserver` (also initialized at the beginning) and a `completedObserver`, reset every time on `reset`. The `positionObserver` is reset on `dispose`.

**Pull request merged**